src/Core/Vipr/FileWriter.cs

```diff
@@ -1,33 +1,89 @@
+using Nito.AsyncEx;
 using System;
 using System.Collections.Generic;
+using System.Collections.Concurrent;
 using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
 using Vipr.Core;
+using NLog;
 
 namespace Vipr
 {
     internal static class FileWriter
     {
-        public static void Write(IEnumerable<TextFile> textFilesToWrite, string outputDirectoryPath = null)
+        private static ConcurrentDictionary<string, AsyncLock> lockDictionary = new ConcurrentDictionary<string, AsyncLock>();
+        internal static Logger Logger => LogManager.GetLogger("FileWriter");
+
+        public static void WriteAsync(IEnumerable<TextFile> textFilesToWrite, string outputDirectoryPath = null)
         {
             if (!string.IsNullOrWhiteSpace(outputDirectoryPath) && !Directory.Exists(outputDirectoryPath))
                 Directory.CreateDirectory(outputDirectoryPath);
 
-            foreach (var file in textFilesToWrite)
-            {
-                var filePath = file.RelativePath;
-
-                if (!string.IsNullOrWhiteSpace(outputDirectoryPath))
+            /* Write asynchronously in batches of 50
+            * This prevents I/O from slowing down
+            * from an overload of requests
+            */
+            var batchSize = 50;

```

Comment:
Possible option would be to extract the batch sizes and pull defaults from constants, if not supplied, but allow for the invoking process to specify the batch sizes. Either by Config file or parameters. Not required ATM, but we could put a TODO to extract the values.